File: registration_system_backend_fe/src/services/team.ts
import { http } from '@/utils/request'

export interface Team {
  id: string
  name: string
  description: string | null
  logo_url: string | null
  captain_id: number | null
  status: number
  credit_score: number
  vip_until: string | null
  trust_label: string
  is_vip: boolean
}

export interface TeamSummary extends Team {
  member_count: number
}

export interface TeamMemberWithInfo {
  user_id: number
  role: string
  role_label: string
  jersey_number: string | null
  joined_at: string
  nickname: string
  real_name: string
  avatar_url: string
  phone_number: string
}

export interface TeamAdminInfo {
  admin_id: number
  username: string
  nickname: string
}

export interface TeamDetailForAdmin {
  team: Team
  members: TeamMemberWithInfo[]
  member_count: number
  assigned_admins: TeamAdminInfo[]
}

export interface TeamCreditTransaction {
  id: number
  team_id: string
  activity_id: string | null
  transaction_type: string
  delta: number
  score_before: number
  score_after: number
  rating: number | null
  amount: string | null
  membership_months: number | null
  note: string | null
  reviewer_team_id: string | null
  created_by_user_id: number | null
  created_by_admin_id: number | null
  created_at: string
}

export interface TeamCreditOverview {
  team: Team
  trust_label: string
  is_vip: boolean
}

export interface CreateTeamPayload {
  name: string
  description?: string
  logo_url?: string
  join_password?: string
  captain_id?: number
}

export interface UpdateTeamPayload {
  name?: string
  description?: string | null
  logo_url?: string | null
  captain_id?: number | null
  status?: number
  join_password?: string | null
}

export interface AddMemberPayload {
  user_id: number
  role?: string
  jersey_number?: string
}

export interface UpdateMemberPayload {
  role?: string
  jersey_number?: string | null
}

export const listTeams = (activeOnly = false) =>
  http.get<TeamSummary[]>('/teams', { params: activeOnly ? { status: 'active' } : {} })

/** 管理后台专用列表（有权限过滤：超级管理员看全部，普通管理员只看被分配的球队） */
export const adminListTeams = (activeOnly = false) =>
  http.get<TeamSummary[]>('/teams/admin-list', { params: activeOnly ? { status: 'active' } : {} })

export const assignAdmin = (teamId: string, adminId: number) =>
  http.post<void>(`/teams/${teamId}/admin-managers`, { admin_id: adminId })

export const unassignAdmin = (teamId: string, adminId: number) =>
  http.delete<void>(`/teams/${teamId}/admin-managers/${adminId}`)

export const listTeamAdmins = (teamId: string) =>
  http.get<TeamAdminInfo[]>(`/teams/${teamId}/admin-managers`)

export const getTeamAdminDetail = (teamId: string) =>
  http.get<TeamDetailForAdmin>(`/teams/${teamId}/admin-detail`)

export const getTeamCreditTransactions = (teamId: string, limit = 20) =>
  http.get<TeamCreditTransaction[]>(`/teams/${teamId}/credit/transactions`, {
    params: { limit },
  })

export const createTeam = (data: CreateTeamPayload) => http.post<Team>('/teams/admin', data)

export const updateTeam = (teamId: string, data: UpdateTeamPayload) =>
  http.patch<void>(`/teams/${teamId}`, data)

export const deleteTeam = (teamId: string) => http.delete<void>(`/teams/${teamId}`)

export const addMember = (teamId: string, data: AddMemberPayload) =>
  http.post<void>(`/teams/${teamId}/members`, data)

export const updateMember = (teamId: string, userId: number, data: UpdateMemberPayload) =>
  http.patch<void>(`/teams/${teamId}/members/${userId}`, data)

export const removeMember = (teamId: string, userId: number) =>
  http.delete<void>(`/teams/${teamId}/members/${userId}`)

export const batchRemoveMembers = (teamId: string, userIds: number[]) =>
  http.delete<number>(`/teams/${teamId}/members/batch`, { data: { user_ids: userIds } })

export const batchUpdateMemberStatus = (teamId: string, userIds: number[], status: number) =>
  http.patch<number>(`/teams/${teamId}/members/batch`, { user_ids: userIds, status })

export const applyTeamCreditPenalty = (teamId: string, points: number, reason: string) =>
  http.post<TeamCreditOverview>(`/teams/${teamId}/credit/penalties`, { points, reason })

File: registration_system_rs/src/team/domain/team.rs
use super::credit::credit_label;
use chrono::NaiveDateTime;
use rust_decimal::Decimal;

/// repository `update` 使用的部分更新字段（None = 不修改）
#[derive(Debug, Default)]
pub struct UpdateTeamFields<'a> {
    pub name: Option<&'a str>,
    pub description: Option<Option<&'a str>>,
    pub logo_url: Option<Option<&'a str>>,
    pub captain_id: Option<Option<i64>>,
    pub status: Option<i8>,
    pub join_password_hash: Option<Option<&'a str>>,
}

#[derive(Debug, Clone)]
pub struct Team {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub logo_url: Option<String>,
    pub captain_id: Option<i64>,
    pub join_password_hash: Option<String>,
    pub status: i8,
    pub credit_score: i32,
    pub vip_until: Option<NaiveDateTime>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

impl Team {
    pub fn is_vip_at(&self, now: NaiveDateTime) -> bool {
        self.vip_until.is_some_and(|vip_until| vip_until >= now)
    }

    pub fn trust_label_at(&self, now: NaiveDateTime) -> String {
        credit_label(self.credit_score, self.is_vip_at(now))
    }
}

#[derive(Debug, Clone)]
pub struct TeamMember {
    pub id: i64,
    pub team_id: i64,
    pub user_id: i64,
    pub role: String,
    pub jersey_number: Option<String>,
    pub joined_at: NaiveDateTime,
    pub status: i8,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone)]
pub struct TeamMemberAttendanceRecord {
    pub activity_id: String,
    pub activity_name: String,
    pub holding_date: NaiveDateTime,
    pub location: String,
    pub stand: i8,
    pub registration_count: i32,
    pub operation_time: Option<NaiveDateTime>,
    pub registered: bool,
}

#[derive(Debug, Clone)]
pub struct TeamAttendanceRankingItem {
    pub user_id: i64,
    pub user_name: String,
    pub avatar_url: Option<String>,
    pub total_count: i64,
    pub attended_count: i64,
    pub leave_count: i64,
    pub late_count: i64,
    pub unregistered_count: i64,
}

/// 球队后台管理员信息
#[derive(Debug, Clone)]
pub struct TeamAdminInfo {
    pub admin_id: i64,
    pub username: String,
    pub nickname: String,
}

/// 队员 + 球员基本信息（管理后台用）
#[derive(Debug, Clone)]
pub struct TeamMemberWithInfo {
    pub user_id: i64,
    pub role: String,
    pub jersey_number: Option<String>,
    pub joined_at: NaiveDateTime,
    pub nickname: String,
    pub real_name: String,
    pub avatar_url: String,
    pub phone_number: String,
}

#[derive(Debug, Clone)]
pub struct TeamCreditTransaction {
    pub id: i64,
    pub team_id: i64,
    pub activity_id: Option<String>,
    pub transaction_type: String,
    pub delta: i32,
    pub score_before: i32,
    pub score_after: i32,
    pub rating: Option<i8>,
    pub amount: Option<Decimal>,
    pub membership_months: Option<i32>,
    pub note: Option<String>,
    pub reviewer_team_id: Option<i64>,
    pub created_by_user_id: Option<i64>,
    pub created_by_admin_id: Option<i64>,
    pub created_at: NaiveDateTime,
}

#[derive(Debug, Clone)]
pub struct ActivityTeamReview {
    pub id: i64,
    pub activity_id: String,
    pub reviewer_team_id: i64,
    pub reviewer_user_id: i64,
    pub reviewee_team_id: i64,
    pub rating: i8,
    pub credit_delta: i32,
    pub comment: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}
